### PR TITLE
mediarecorder: prefer VP9 when possible and add mp4 fallback

### PR DESCRIPTION
**Pull request by GarboMuffin**
Resolves https://github.com/ScratchAddons/ScratchAddons/issues/7055 (lots of discussion in there), based on changes in #7006

This logic is in this order:

1. Try to record as webm using VP9 (Chrome, best scenario)
2. Try to record as webm using whatever codec the browser gives us (Firefox will give us VP8)
3. Try to record as mp4 using whatever codec the browser gives us (Safari will give us H264)

Here is a video recorded by Chromium on Linux: https://github.com/ScratchAddons/ScratchAddons/assets/33787854/6e7ec5f3-aaab-4e25-94ec-5025bff8fc05 - mpv says this is vp9

Here is a video recorded by Firefox on Linux: https://github.com/ScratchAddons/ScratchAddons/assets/33787854/f4fb588a-08a0-4299-9dbd-0229c0686e57 - mpv says this is vp8

Here is a video recorded by Safari: https://github.com/ScratchAddons/ScratchAddons/assets/33787854/6d180f9e-117c-4b65-9b22-0c5ddca90ed2 - mpv says this is h264

### CODE REVIEW

**Comment by GarboMuffin** [comment]
@Joeclinton1 does this work for you?

**Comment by Joeclinton1** [comment]
> @Joeclinton1 does this work for you?

The code looks good. The .find() is an elegant approach, I tested on Chrome (windows) and it works. The video exported to VP9 as expected. I can't figure out how to load the extension in firefox so couldn't test that. I also can't test Safari. I expect if it works for you that's a good sign it will work for others.

Assuming isTypeSupported works as we think it does, I can't think of any edge cases where this logic would break.

**Comment by Joeclinton1** [comment]
I have tested in Firefox now and it exports to vp8 as expected. The vp8 video can be viewed in discord, although it lacks the scrubbing functionality that the vp9 has.

**Review by Joeclinton1 — APPROVED** [review]
Looks good to me!

**Comment by WorldLanguages** [comment]
Nice work guys!

**Comment by DNin01** [comment]
Totally works (including the microphone) on iOS Safari in TurboWarp!

File: addons/mediarecorder/userscript.js [hunk 1]
```diff
@@ -10,6 +10,18 @@ export default async ({ addon, console, msg }) => {
...
   let timeout;
+
+  const mimeType = [
+    // Chrome and Firefox only support encoding as webm
+    // VP9 is preferred as its playback is better supported across platforms
+    "video/webm; codecs=vp9",
+    // Firefox only supports encoding VP8
+    "video/webm",
+    // Safari only supports encoding H264 as mp4
+    "video/mp4",
```
Comment (WorldLanguages) [inline addons/mediarecorder/userscript.js:21]: As of Chrome 126, `video/mp4` is supported: https://chromestatus.com/feature/5163469011943424

@GarboMuffin Should we move it to the top of this list?

Comment (GarboMuffin) [reply addons/mediarecorder/userscript.js:21]: i think that would make a lot of people happier

**Comment by WorldLanguages** [comment]
Feedback from a Chrome user:

> please change the video output type back into .webm. it doesnt even have an audio anymore despite turning the "record audio" option on, and it cant even be played on most media player softwares

**Comment by DNin01** [comment]
> Feedback from a Chrome user:
>
> > please change the video output type back into .webm. it doesnt even have an audio anymore despite turning the "record audio" option on, and it cant even be played on most media player softwares

Can replicate with Edge 131 on Windows 11. Maybe we should revert that change for now? And then we can bring it back when this issue is resolved.

**Comment by DNin01** [comment]
I won't mention mp4 support on my next blog post.

**Comment by WorldLanguages** [comment]
Here is my PR #8022

**Comment by Samq64** [comment]
I can't reproduce on Chromium 131 on Linux. It's the project audio and not the microphone, right?

**Comment by DNin01** [comment]
> I can't reproduce on Chromium 131 on Linux. It's the project audio and not the microphone, right?

Yes, the project audio isn't being recorded. I didn't test the microphone.

**Comment by Joeclinton1** [comment]
I guess I never tested audio, I didn't even know we recorded audio
